rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +921,129 @@ void PlayerImpl::load_storage_content()
+    const bool no_messages = std::all_of(
+      input_bags_.cbegin(),
+      input_bags_.cend(),
+      [](const auto & reader_options) {return !reader_options.first->has_next();});
+    if (no_messages) {
+      break;
+    }
 
-    if (message_queue_.size_approx() < queue_lower_boundary) {
-      enqueue_up_to_boundary(queue_upper_boundary);
+    // The message queue size may get smaller after this, but that's OK
+    const size_t message_queue_size = message_queue_.size();
+    if (message_queue_size < queue_lower_boundary) {
+      enqueue_up_to_boundary(queue_upper_boundary, message_queue_size);
     } else {
       lk.unlock();
       std::this_thread::sleep_for(std::chrono::milliseconds(1));
     }
   }
 }
 
-void PlayerImpl::enqueue_up_to_boundary(size_t boundary)
+void PlayerImpl::enqueue_up_to_boundary(const size_t boundary, const size_t message_queue_size)
 {
-  rosbag2_storage::SerializedBagMessageSharedPtr message;
-  for (size_t i = message_queue_.size_approx(); i < boundary; i++) {
-    if (!reader_->has_next()) {
-      break;
+  // Read messages from input bags in a round robin way
+  size_t input_bag_index = 0u;
+  for (size_t i = message_queue_size; i < boundary; i++) {
+    const auto & reader = input_bags_[input_bag_index].first;
+    // We are supposed to have at least one bag with messages to read
+    if (reader->has_next()) {
+      message_queue_.push(reader->read_next());
     }
-    message = reader_->read_next();
-    message_queue_.enqueue(message);
+    input_bag_index = (input_bag_index + 1) % input_bags_.size();
   }
 }
 
 void PlayerImpl::play_messages_from_queue()
 {
-  // Note: We need to use message_queue_.peek() instead of message_queue_.try_dequeue(message)
-  // to support play_next() API logic.
-  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = peek_next_message_from_queue();
-  { // Notify play_next() that we are ready for playback
-    // Note: We should do notification that we are ready for playback after peeking pointer to
-    // the next message. message_queue_.peek() is not allowed to be called from more than one
-    // thread concurrently.
+  { // Notify play_next()/seek() that we are ready for playback
     std::lock_guard<std::mutex> lk(ready_to_play_from_queue_mutex_);
     is_ready_to_play_from_queue_ = true;
     ready_to_play_from_queue_cv_.notify_all();
   }
-  while (rclcpp::ok() && !stop_playback_ &&
-    message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
-  {
-    // Do not move on until sleep_until returns true
-    // It will always sleep, so this is not a tight busy loop on pause
-    while (rclcpp::ok() && !clock_->sleep_until(message_ptr->recv_timestamp)) {
-      if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
-        break;
+  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = take_next_message_from_queue();
+
+  // While we haven't stopped playing
+  while (rclcpp::ok() && !stop_playback_) {
```

Comment:
I think it does make sense to add a comment that we need this outer loop for the cases when playback in pause i.e. need to wait on `clock.sleep_until(..)` and for the play next operation

Reply:
added in f26ee05b2533fa263ef2c35784081f540f4f6b8e